processor/src/main/java/org/mapstruct/ap/internal/conversion/JavaStreamToCollectionConversion.java

```
@@ -0,0 +1,95 @@
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import org.mapstruct.ap.internal.model.common.ConversionContext;
+import org.mapstruct.ap.internal.model.common.Type;
+import org.mapstruct.ap.internal.model.common.TypeFactory;
+import org.mapstruct.ap.internal.util.Collections;
+import org.mapstruct.ap.internal.util.JavaStreamConstants;
+
+/**
+ * Conversion between {@link java.util.stream.Stream} and java utils: {@link Set}, {@link java.util.List},
+ * {@link java.util.Collection}.
+ *
+ * @author Filip Hrisafov
+ */
+public class JavaStreamToCollectionConversion extends SimpleConversion {
+
+    @Override
+    protected String getToExpression(ConversionContext conversionContext) {
+        return "Optional.ofNullable( <SOURCE> )" +
+            ".map( s -> s.collect( Collectors." + collector( conversionContext ) + " ) )" +
+            ".orElse( null )";
+    }
+
+    private String collector(ConversionContext conversionContext) {
+        Type targetType = conversionContext.getTargetType();
+        TypeFactory typeFactory = conversionContext.getTypeFactory();
+        String collector = "";
+        // we must do erasure as JDK compiler does not work without it
+        //TODO is this a bug in the Java compiler or not?
+        if ( typeFactory.getType( Iterable.class ).isAssignableTo( targetType.erasure() )
+            || typeFactory.getType( Collection.class ).isAssignableTo( targetType.erasure() )
+            || typeFactory.getType( List.class ).isAssignableTo( targetType.erasure() ) ) {
+            collector = "toList()";
+        }
+        else if ( typeFactory.getType( Set.class ).isAssignableTo( targetType.erasure() ) ) {
+            collector = "toSet()";
+        }
+        else if ( acceptsTreeSet( conversionContext ) ) {
+            collector = "toCollection( TreeSet::new )";
+        }
+
+        //TODO what if we register wrong. In the end it won't compile, but should we throw an exception?
```

Yes, we should never generate code that doesn't compile. Instead we should raise a descriptive error about the issue before (missing conversion etc.).